### ISSUE TITLE
No drop flag check for DST structs

### ISSUE THREAD

**Issue by nrc** [rust-lang/rust#16758]
Since drop flags come after all named fields and the last field in a DST struct is unsized, finding the drop flag is a bit tricky. Therefore, it is currently not checked. A possible fix is to find the alignment by taking the size of the struct from its vtable and subtracting the size of the drop flag. However, since the plan is to remove the drop flag anyway (#5016), I have just left a FIXME and issue a warning when we skip the check.

cc @pnkfelix

**Comment by steveklabnik** [comment]
@nrc is this still relevant today?

**Comment by nrc** [comment]
Yes, afaik. cc @pnkfelix

**Comment by tailhook** [comment]
Does this mean if i get:
```
warning: Ignoring drop flag in destructor for std::sync::mutex::Mutex<Something>because the struct is unsized. See issue#16758
```
Then I will get a memory leak?

**Comment by pnkfelix** [comment]
some relevant recent discussion from `#rust-internals` IRC channel: https://botbot.me/mozilla/rust-internals/2016-04-28/?msg=65078367&page=3

**Comment by chriscoomber** [comment]
Just hit this when coercing a `Box<RwLock< (struct) >>` to a `Box<RwLock< (trait) >>`. 

(Aside: it is of course illegal to coerce a `RwLock< (struct) >` to a `RwLock< (trait) >` - I don't really understand why the outer box makes this legal.)

Here's a small example that produces this warning:

```rust
use std::sync::RwLock;
use std::boxed::Box;

trait Trait {}
struct Thing;
impl Trait for Thing {}

fn main() {
  let _: Box<RwLock<Trait>> = Box::new(RwLock::new(Thing));
}
```

Will I get a memory leak with this code?

**Comment by jonas-schievink** [comment]
This should no longer be relevant after #35764 lands since it will remove in-struct drop flags

**Comment by Osspial** [comment]
Should this warning be removed now? Non-zeroing drops have landed, so there isn't a flag to check.

**Comment by arielb1** [comment]
Sure enough. There is no such a thing as a drop flag in the MIR (1.13+) universe.
